sphericart-jax/src/sphericart_jax_cuda.cpp

```diff
@@ -2,38 +2,116 @@
 // devices. It is exposed as a standard pybind11 module defining "capsule"
 // objects containing our methods. For simplicity, we export a separate capsule
 // for each supported dtype.
-// This file is separated from `sphericart_jax_cuda.cu` because pybind11 does
-// not accept cuda files.
 
 #include <cstdlib>
 #include <map>
 #include <mutex>
 #include <tuple>
 
+#include "sphericart_cuda.hpp"
 #include "sphericart/pybind11_kernel_helpers.hpp"
-#include "sphericart/sphericart_jax_cuda.hpp"
 
-using namespace sphericart_jax;
-using namespace sphericart_jax::cuda;
+struct SphDescriptor {
+    std::int64_t n_samples;
+    std::int64_t lmax;
+};
 
-namespace {
+namespace sphericart_jax {
+namespace cuda {
 
-// Registration of the custom calls with pybind11
+template <template <typename> class C, typename T>
+using CacheMapCUDA = std::map<size_t, std::unique_ptr<C<T>>>;
+
+template <template <typename> class C, typename T>
+std::unique_ptr<C<T>>& _get_or_create_sph_cuda(size_t l_max) {
+    // Static map to cache instances based on parameters
+    static CacheMapCUDA<C, T> sph_cache;
+    static std::mutex cache_mutex;
+
+    // Check if instance exists in cache, if not create and store it
```

Comment:
@frostedoyster should we purge the instances after some point? Or do we let this map grow unbounded? (This would be for a separate PR!)